firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java

```diff
@@ -87,6 +108,259 @@ public void addFieldIndex(FieldIndex index) {
+                      documentPath,
+                      fieldIndex);
+                }
+
+                List<byte[]> encodeValues = encodeDocumentValues(fieldIndex, values);
+                for (byte[] encoded : encodeValues) {
+                  // TODO(indexing): Handle different values for different users
+                  db.execute(
+                      "INSERT OR IGNORE INTO index_entries ("
+                          + "index_id, "
+                          + "index_value, "
+                          + "document_name) VALUES(?, ?, ?)",
+                      indexId,
+                      encoded,
+                      documentPath.canonicalString());
+                }
+              } catch (InvalidProtocolBufferException e) {
+                throw fail("Invalid index: " + e);
+              }
+            });
+  }
+
+  /**
+   * Returns the list of values for all fields indexed by the provided field index. Returns {@code
+   * null} if one or more of the fields are not set.
+   */
+  @Nullable
+  private List<Value> extractFieldValue(Document document, FieldIndex fieldIndex) {
+    List<Value> values = new ArrayList<>();
+    for (FieldIndex.Segment segment : fieldIndex) {
+      Value field = document.getField(segment.getFieldPath());
+      if (field == null) {
+        return null;
+      }
+      values.add(field);
+    }
+    return values;
+  }
+
+  @Override
+  @Nullable
+  public Set<DocumentKey> getDocumentsMatchingTarget(Target target) {
+    @Nullable FieldIndex fieldIndex = getMatchingIndex(target);
+    if (fieldIndex == null) return null;
+
+    Bound lowerBound = target.getLowerBound(fieldIndex);
+    @Nullable Bound upperBound = target.getUpperBound(fieldIndex);
+
+    if (Logger.isDebugEnabled()) {
+      Logger.warn(
+          TAG,
+          "Using index '%s' to execute '%s' (Lower bound: %s, Upper bound: %s)",
+          fieldIndex,
+          target,
+          lowerBound,
+          upperBound);
+    }
+
+    Set<DocumentKey> result = new HashSet<>();
+
+    if (upperBound != null) {
+      List<byte[]> lowerBoundValues =
+          encodeTargetValues(fieldIndex, target, lowerBound.getPosition());
+      List<byte[]> upperBoundValues =
+          encodeTargetValues(fieldIndex, target, upperBound.getPosition());
+
+      hardAssert(
+          lowerBoundValues.size() == upperBoundValues.size(),
+          "Expected upper and lower bound size to match");
+
+      // TODO(indexing): To avoid reading the same documents multiple times, we should ideally only
+      // send one query that combines all clauses.
+      for (int i = 0; i < lowerBoundValues.size(); ++i) {
+        db.query(
+                String.format(
+                    "SELECT document_name from index_entries WHERE index_id = ? AND index_value %s ? AND index_value %s ?",
+                    lowerBound.isBefore() ? ">=" : ">", upperBound.isBefore() ? "<=" : "<"))
```

Comment:
Should the order of operator for upper bound be reversed? 

I am always confused with `isBefore()`, so I could be wrong. But if I am right, please add a test for this.

Reply:
Actually there are tests for this, so the code is right I guess. Can you please add some comment to explain?

Reply:
I myself am confused about this all the time. It takes a long time to understand - and it looks like I got it wrong here as well. The tests were passing because I invalidated invalid assumptions. This was obvious when I added a small comment, which strongly hinted that I was doing something wrong.